mcompile: replaced backend divination code + cleanup

TheQwertiest opened the pull request:
- Replaced backend divination code with coredata parsing (`meson compile` should be used only with meson configured folders anyway).
- Extracted backend specific code to separate functions.
- Improved(?) error message when invoked on non-configured directory.

TheQwertiest commented:
CI failures seem unrelated:
```
Failed test during configure: 'test cases/warning/1 version for string div'
```

xclaesse commented on mesonbuild/mcompile.py:
```diff
@@ -27,6 +28,20 @@
+    """
+    Gets `backend` option value from introspection data
+    """
+    path_to_intro = builddir / 'meson-info' / 'intro-buildoptions.json'
```
I think you should get it from coredata file instead. Should be faster to pickle that file than parsing json.

TheQwertiest replied:
Fixed

TheQwertiest commented:
Changes in the latest commit:
- Replaced backend divination code with coredata parsing instead of introspect.
- Extracted backend specific code to separate functions.
- Improved(?) error message when invoked on non-configured directory.

xclaesse commented on mesonbuild/mcompile.py:
```diff
@@ -15,19 +15,80 @@
+    return coredata.load(str(builddir)).get_builtin_option('backend')
+
+def get_parsed_args_ninja(options: 'argparse.Namespace', builddir: Path):
+    runner = os.environ.get('NINJA')
```
Better use detect_ninja() from environment.py here.

TheQwertiest replied:
Fixed

xclaesse commented on mesonbuild/mcompile.py:
```diff
@@ -15,19 +15,80 @@
+    if (not builddir.is_dir() or
+        not (builddir / 'meson-private').is_dir() or
+        not (builddir / 'meson-info').is_dir() or
+        not (builddir / 'meson-info' / 'meson-info.json').is_file()):
```
Why do you check each subdir? Can't you just check if coredata.dat exists?

TheQwertiest replied:
Haha, fair enough (I've just blindly copy pasted this check from mintro)

xclaesse replied:
Right, my 2 remaining nitpickings are not your fault, was already like that, but while you're here it would be nice to cleanup :)

TheQwertiest replied:
I like to leave the code in the better state than I've found it, so all nit-picks are appreciated.

TheQwertiest replied:
Fixed

xclaesse commented:
Looks good to me, thanks.
